burstphoto/align.metal

```diff
@@ -94,31 +95,200 @@ kernel void avg_pool(texture2d<float, access::read> in_texture [[texture(0)]],
-                                 texture2d<uint, access::write> out_texture [[texture(1)]],
-                                 constant int& divisor [[buffer(0)]],
-                                 uint2 gid [[thread_position_in_grid]]) {
-    float in_color = in_texture.read(gid).r;
-    uint out = int(round(in_color * UINT16_MAX_VAL / divisor));
-    out_texture.write(out, gid);
+kernel void blur_mosaic_x(texture2d<float, access::read> in_texture [[texture(0)]],
+                          texture2d<float, access::write> out_texture [[texture(1)]],
+                          constant int& kernel_size [[buffer(0)]],
+                          constant int& mosaic_pattern_width [[buffer(1)]],
+                          uint2 gid [[thread_position_in_grid]]) {
+    
+    // load args
+    int texture_width = in_texture.get_width();
+    
+    // clamp kernel_size to a maximum of 8
+    int const kernel_size_c = clamp(kernel_size, 0, 8);
+    
+    // set kernel weights of binomial filter
+    float bw[9] = {1, 0, 0, 0, 0, 0, 0, 0, 0};
+    
+    if(kernel_size_c == 1)      {bw[0]=    2; bw[1]=    1;}
+    else if(kernel_size_c == 2) {bw[0]=    6; bw[1]=    4; bw[2]=   1;}
+    else if(kernel_size_c == 3) {bw[0]=   20; bw[1]=   15; bw[2]=   6; bw[3]=   1;}
+    else if(kernel_size_c == 4) {bw[0]=   70; bw[1]=   56; bw[2]=  28; bw[3]=   8; bw[4]=   1;}
+    else if(kernel_size_c == 5) {bw[0]=  252; bw[1]=  210; bw[2]= 120; bw[3]=  45; bw[4]=  10; bw[5]=  1;}
+    else if(kernel_size_c == 6) {bw[0]=  924; bw[1]=  792; bw[2]= 495; bw[3]= 220; bw[4]=  66; bw[5]= 12; bw[6]=  1;}
+    else if(kernel_size_c == 7) {bw[0]= 3432; bw[1]= 3003; bw[2]=2002; bw[3]=1001; bw[4]= 364; bw[5]= 91; bw[6]= 14; bw[7]= 1;}
+    else if(kernel_size_c == 8) {bw[0]=12870; bw[1]=11440; bw[2]=8008; bw[3]=4368; bw[4]=1820; bw[5]=560; bw[6]=120; bw[7]=16; bw[8]=1;}
+
+    // compute a sigle output pixel
+    float total_intensity = 0;
+    float total_weight = 0;
+    int y = gid.y;
+    
+    for (int dx = -kernel_size_c; dx <= kernel_size_c; dx++) {
+        int x = gid.x + mosaic_pattern_width*dx;
+        if (0 <= x && x < texture_width) {
+
+            float const weight = bw[abs(dx)];
+            total_intensity += weight * in_texture.read(uint2(x, y)).r;
+            total_weight += weight;
+        }
+    }
+    
+    // write output pixel
+    float out_intensity = total_intensity / total_weight;
+    out_texture.write(out_intensity, gid);
 }
 
 
+kernel void blur_mosaic_y(texture2d<float, access::read> in_texture [[texture(0)]],
+                          texture2d<float, access::write> out_texture [[texture(1)]],
+                          constant int& kernel_size [[buffer(0)]],
+                          constant int& mosaic_pattern_width [[buffer(1)]],
+                          uint2 gid [[thread_position_in_grid]]) {
+    
+    // load args
+    int texture_height = in_texture.get_height();
```

Comment:
Why does this have to be calculated inside the kernel? Can't this be passed as a parameter?

Reply:
I think Metal is extremely efficient in extracting this information from the texture. Still it can be checked in a future version if performance can be increased.

Reply:
This suggestion does not show any improvement of performance.